### PR TITLE
Feature/24 portable map widget

### PR DESCRIPTION
**Pull request by cschwinderg**
## Related Issues:
* [HMW-24](https://jira.epa.gov/browse/HMW-24)

## Main Changes:
* Added a save panel to the Add data widget, which saves HMW data to ArcGIS Online.

## Steps To Test:
1. Navigate to http://localhost:3000/community/dc/overview
2. Select layers to save 
3. Click save and login
4. Go to your  AGO account and verify you have the expected output

## TODO:
I'll start working on this stuff, but I'm ok with merging this in without completing any of the below items. If I don't get to these I'll create separate tickets for these items.

- [x] Add publishing logic for the Upstream Watershed layer
- [x] Add publishing logic for the Identified Issues layer
- [ ] Add publishing logic for the CyAN layer(s)
- [x] Update the `isServiceNameAvailable` function to check for uniqueness in the user's content instead of at the organization level. 
    *  I confirmed in the esri docs that `isServiceNameAvailable`  checks for uniqueness across the organization on not just in the user's content. 
    * I did some more testing on this and esri just doesn't allow duplicate names across the org. So this is item just isn't possible and I'm going to go ahead and mark this one off.
- [ ] Update the HMW generated portion of the description to say what filters have been applied to layers, if any (i.e., dischargers with significant violations, identified issues, etc.).
- [x] ~Fix bug related to saving layers uploaded from a file where they have the same file name.~ Fix bug related to saving multiple layers uploaded from a file where the file name ends with `.zip`.
- [ ] Update popups to match up with HMW more

### CODE REVIEW

**Comment by cschwinderg** [comment]
@maxdiebold-erg I just realized that I forgot to mention that you will need a client id environment variable for the AGO login to work. I sent a separate email that includes the client id that you will need to test this PR. I'm sorry for any confusion.

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+  const {
+    saveAsName,
+    setSaveAsName,
+    saveDescription,
+    setSaveDescription,
+    saveLayersList,
+    setSaveLayersList,
+    widgetLayers,
+  } = useAddSaveDataWidgetState();
+  const mapView = useContext(LocationSearchContext).mapView as __esri.MapView;
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:195]: This type cast is iffy, because `mapView` could technically be `null`, but I think in this case it's fine, since there's no way to open the widget while the map is loading.

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+        if (
+          layersToIgnore.includes(layer.id) ||
+          layerTypesToIgnore.includes(layer.type)
+        )
+          return;
+
+        if (newSwitches.hasOwnProperty(layer.id)) {
+          newSwitches[layer.id].layer = layer;
+        }
+      });
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:319]: Currently, this block doesn't change state. Is `saveLayersList` supposed to be updated here?
```suggestion
      });
      setSaveLayersList(newSwitches);
```

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+      let totalFeatures = 0;
+      let featureFailedCount = 0;
+      response.featuresResult?.forEach((l) => {
+        l.addResults?.forEach((r) => {
+          totalFeatures += 1;
+          if (!r.success) featureFailedCount += 1;
+        });
+      });
+
+      const webMapSuccess = response.webMapResult.success;
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:463]: If the _addItem_ endpoint is blocked (or fails), `webMapResult` is undefined and this throws a type error, so might want to guard against that here.
```suggestion
      const webMapSuccess = response.webMapResult?.success ?? false;
```

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+        });
+      } else {
+        let errorMessage = '';
+
+        if (!layerSuccess)
+          errorMessage +=
+            'Failed to save 1 or more layers. Check console for more information.\n';
+        if (featureFailedCount > 0)
+          errorMessage += `Partial save success. Failed to save ${featureFailedCount} of ${totalFeatures} feature(s).\n`;
+        if (webMapSuccess) errorMessage += 'Failed to save the web map.\n';
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:477]: It looks like this should be negated:
```suggestion
        if (!webMapSuccess) errorMessage += 'Failed to save the web map.\n';
```

**Review by maxdiebold-erg — CHANGES_REQUESTED** [review]
This is really cool! It seems to work well from what I've tested so far. I'll keep testing it out, but I noted a few initial things I found below.

Additionally, I saw that the `locationUrl` for the monitoring locations just has the relative path that we use in HMW. Then, I looked at the _layerProps_ config and it seems like that attribute shouldn't have a value at all (length -1), so I'm not how it got included in the saved map.

**Comment by maxdiebold-erg** [comment]
Also, I noticed while looking at the attributes of a _Past Water Conditions_ feature that the `station[xxx]` attributes are blank. This is my fault, I truncated the attributes name when centralizing the `monitoringLocationsLayer`, and removed "station" from some of the attribute names, I didn't think it benefited the attribute description. So we'll need to update the _layerProps_ or the layer itself.

**Comment by cschwinderg** [comment]
> Additionally, I saw that the `locationUrl` for the monitoring locations just has the relative path that we use in HMW. Then, I looked at the _layerProps_ config and it seems like that attribute shouldn't have a value at all (length -1), so I'm not sure how it got included in the saved map.

The `length: -1` in AGO basically means max possible length. I'm probably going to change that to `length: 4000`, because in AGO you just see something like `locationUrl: View (link)` instead of `locationUrl: https://some-path.gov....` when -1 is used for length.

**Comment by cschwinderg** [comment]
> Also, I noticed while looking at the attributes of a _Past Water Conditions_ feature that the `station[xxx]` attributes are blank. This is my fault, I truncated the attributes name when centralizing the `monitoringLocationsLayer`, and removed "station" from some of the attribute names, I didn't think it benefited the attribute description. So we'll need to update the _layerProps_ or the layer itself. I'm not sure if it's necessary for this pull request, though.

Thank you for bringing that to my attention. I'll fix go ahead and fix that.

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+
+const tooltipCost =
+  'Including this layer may require ArcGIS Online credits for storage.';
+const tooltipFiltered = 'This layer will be saved with your selected filters.';
+const tooltipNotLoaded = 'This layer may not have been loaded yet.';
+
+const layersToIgnore = [
+  'nonprofitsLayer',
+  'protectedAreasHighlightLayer',
+  'surroundingMonitoringLocationsLayer',
```
[the remark this answers is not in the document]
Comment (cschwinderg) [reply app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:50]: Good catch! I'll remove that one.

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx (hunk 1 above)
Comment (cschwinderg) [reply app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:195]: I didn't really like doing this. I just did it to get some kind of type, even though the LocationSearchContext isn't typescript ready yet. Once we get the LocationSearchContext updated for TypeScript we can get rid of this code.

I will go ahead and change this to `const mapView = useContext(LocationSearchContext).mapView as __esri.MapView | null;` so the type is atleast accurate.

File: app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx (hunk 2 above)
Comment (cschwinderg) [reply app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx:319]: Good catch! That entire `useEffect` is unnecessary, so I'm going to pull that code out. I think that was some code that was mainly for testing, that I just forgot to remove.

**Comment by cschwinderg** [comment]
@maxdiebold-erg I'm done making changes to this for now. Can you re-review this PR?

I did go ahead and update the Past Water Conditions such that the locationUrl attribute contains the full URL. Also the `length: -1` does not make AGO display `locationUrl: View (link)`. That just happens when AGO determines that the value is a URL.

**Comment by cschwinderg** [comment]
Also, I created tickets for the incomplete TODO items above.

**Review by maxdiebold-erg — APPROVED** [review]
I see, the -1 character limit makes sense. This looks great! There's a lot to it, so I'll let you know if I see anything else, but it should help to get others testing it out, too.

**Comment by cschwinderg** [comment]
Sounds good! I'll merge it in as-is. Any issues we find will be separate tickets.
